cpp/net/url_fetcher.cc

```diff
@@ -0,0 +1,200 @@
+#include "net/url_fetcher.h"
+
+#include <event2/buffer.h>
+#include <event2/keyvalq_struct.h>
+#include <glog/logging.h>
+
+#include "net/connection_pool.h"
+
+using cert_trans::internal::ConnectionPool;
+using cert_trans::internal::evhttp_connection_unique_ptr;
+using std::chrono::seconds;
+using std::chrono::steady_clock;
+using std::lock_guard;
+using std::make_pair;
+using std::map;
+using std::move;
+using std::mutex;
+using std::pair;
+using std::string;
+using std::unique_ptr;
+using util::Status;
+using util::Task;
+using util::TaskHold;
+
+namespace cert_trans {
+
+
+struct UrlFetcher::Impl {
+  Impl(libevent::Base* base) : pool_(CHECK_NOTNULL(base)) {
+  }
+
+  internal::ConnectionPool pool_;
+};
+
+
+namespace {
+
+
+struct State {
+  State(ConnectionPool* pool, evhttp_cmd_type verb,
+        const UrlFetcher::Request& req, UrlFetcher::Response* resp,
+        Task* task);
+
+  ~State() {
+    pool_->Put(move(conn_));
+  }
+
+  ConnectionPool* const pool_;
+  evhttp_connection_unique_ptr conn_;
+  const UrlFetcher::Request request_;
+  UrlFetcher::Response* const response_;
+  Task* const task_;
+
+  evhttp_request* const http_req_;
+};
+
+
+void RequestCallback(evhttp_request* req, void* userdata) {
+  State* const state(static_cast<State*>(CHECK_NOTNULL(userdata)));
+
+  if (!req) {
+    // TODO(pphaneuf): The dreaded null request... These are fairly
+    // fatal things, like protocol parse errors, but could also be a
+    // connection timeout. I think we should do retries in this case,
+    // with a deadline of our own? At least, then, it would be easier
+    // to distinguish between an obscure error, or a more common
+    // timeout.
+    VLOG(1) << "RequestCallback received a null request";
+    state->task_->Return(Status::UNKNOWN);
+    return;
+  }
+
+  state->response_->status_code = evhttp_request_get_response_code(req);
+  if (state->response_->status_code < 100) {
+    // TODO(pphaneuf): According to my reading of libevent, this is
+    // most likely to be a connection refused?
+    VLOG(1) << "request has a status code lower than 100: "
+            << state->response_->status_code;
+    state->task_->Return(
+        Status(util::error::FAILED_PRECONDITION, "connection refused"));
+    return;
+  }
+
+  map<string, string, ci_less<string>> headers;
+  for (evkeyval* ptr = evhttp_request_get_input_headers(req)->tqh_first; ptr;
+       ptr = ptr->next.tqe_next) {
+    map<string, string>::iterator it(headers.find(ptr->key));
+
+    if (it == headers.end()) {
+      headers.insert(make_pair(ptr->key, ptr->value));
+    } else {
+      it->second.append(", ");
```

Comment:
Since C++11 (<3 C++11) the order of values associated with the same key into a multimap is order of insertion (http://en.cppreference.com/w/cpp/container/multimap), so I think you're ok from that PoV, but I'll leave it up to you to decide.